packages/react/src/host/controller.tsx

```diff
@@ -1,16 +1,25 @@
 import type {ComponentType} from 'react';
 import type {Controller, Renderer} from './types';
 
-import {renderComponent} from './RemoteComponent';
-import {renderText} from './RemoteText';
+import {renderComponent as defaultRenderComponent} from './RemoteComponent';
+import {renderText as defaultRenderText} from './RemoteText';
 
 export interface ComponentMapping {
   [key: string]: ComponentType<any>;
 }
 
+interface RendererFactory {
+  componentRenderer: (
+    defaultRenderer: Renderer['renderComponent'],
+  ) => Renderer['renderComponent'];
```

Comment:
Specifically this ⬆️ 

This is how the consumer can use the api

```ts
createController(components, {
  componentRenderer: (defaultRenderer) => (props) => {
    const {component} = props;
    switch(component.type) {
       case 'Card': 
         // Don't render Card
         return null;
       case 'Stack':
         // return custom Stack component
         break;
       default:
          return defaultRenderer(props);
    }
  }
});
```

Reply:
Looking at this with fresh eyes, this API feels a little too complex and verbose. I would prefer this:

```ts
createController(components, {
  renderComponent(component, {key, renderDefault}) {
    switch (component.type) {
      case 'Card': return null;
      case 'Stack': return <CustomStackComponent {...component.props} key={key} />;
      default: return renderDefault();
    }
  }
});
```

We get rid of the "arrow function returning arrow function" business and move some of the arguments around so that `component`/ `text` can be the first positional argument, which lets consumers destructure a little more confidently since they never need to pass the arguments they get back into another function. What do you think?

Reply:
Do you think we need to make the usage of key transparent? The consumer of this renderComponent should not know about why they need the key. If they forget to use key, it will complain because the part that needs it is in RemoteRenderer. Other than that, the approach looks good to me

Reply:
That should be okay. Let's do it @james-woo. Action items: 
- Use `Fragment` for the key
- Change signature to `renderComponent(component, {renderDefault})`. I think we still keep renderDefault in an object for future extensibility.

Reply:
👍